File: generators/checkboard.cpp

// Part of the ProceduralTextureMaker project.
// http://github.com/johanokl/ProceduralTextureMaker
// Released under GPLv3.
// Johan Lindqvist (johan.lindqvist@gmail.com)

#include "checkboard.h"
#include <QColor>
#include <QPainter>

CheckboardTextureGenerator::CheckboardTextureGenerator() {
   TextureGeneratorSetting color;
   color.name = "Colour";
   color.description = "Colour used for alternating squares.";
   color.defaultvalue = QVariant(QColor(0, 0, 0, 255));
   color.id = "color";
   configurables.append(color);

   TextureGeneratorSetting brickwidth;
   brickwidth.name = "Square width";
   brickwidth.description = "Width of each square, relative to a 300-pixel texture width.";
   brickwidth.defaultvalue = QVariant((int)10);
   brickwidth.min = QVariant(1);
   brickwidth.max = QVariant(300);
   brickwidth.id = "brickwidth";
   configurables.append(brickwidth);

   TextureGeneratorSetting brickheight;
   brickheight.name = "Square height";
   brickheight.description = "Height of each square, relative to a 300-pixel texture height.";
   brickheight.defaultvalue = QVariant((int)10);
   brickheight.min = QVariant(1);
   brickheight.max = QVariant(300);
   brickheight.id = "brickheight";
   configurables.append(brickheight);

   TextureGeneratorSetting offsetx;
   offsetx.name = "Horizontal offset (%)";
   offsetx.description = "Moves the pattern horizontally as a percentage of the texture width.";
   offsetx.defaultvalue = QVariant((int)0);
   offsetx.min = QVariant(-100);
   offsetx.max = QVariant(100);
   offsetx.id = "offsetx";
   configurables.append(offsetx);

   TextureGeneratorSetting offsety;
   offsety.name = "Vertical offset (%)";
   offsety.description = "Moves the pattern vertically as a percentage of the texture height.";
   offsety.defaultvalue = QVariant((int)0);
   offsety.min = QVariant(-100);
   offsety.max = QVariant(100);
   offsety.id = "offsety";
   configurables.append(offsety);
}
void CheckboardTextureGenerator::generate(QSize size, TexturePixel* destimage,
                                          const QMap<QString, TextureImagePtr>& sourceimages,
                                          const TextureNodeSettings& settings) const {
   if (!destimage || !size.isValid()) {
      return;
   }
   QColor color = settings.value("color").value<QColor>();
   int brickheight =
       qMax(1, static_cast<int>(settings.value("brickheight").toDouble() * size.height() / 300));
   int brickwidth =
       qMax(1, static_cast<int>(settings.value("brickwidth").toDouble() * size.width() / 300));
   int offsetx = settings.value("offsetx").toDouble() * size.width() / 100;
   int offsety = settings.value("offsety").toDouble() * size.height() / 100;

   if (sourceimages.contains(QStringLiteral("Input"))) {
      memcpy(destimage, sourceimages.value(QStringLiteral("Input"))->getData(),
             size.width() * size.height() * sizeof(TexturePixel));
   } else {
      memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
   }
   QImage tempimage = makeTextureImageView(size, destimage);
   QPainter painter(&tempimage);
   painter.setCompositionMode(QPainter::CompositionMode_SourceOver);
   painter.setBrush(QBrush(color, Qt::BrushStyle::SolidPattern));
   painter.setPen(Qt::NoPen);

   bool invert = false;
   int currY = offsety - size.height() - brickheight;
   while (currY < size.height()) {
      invert = !invert;
      int currX = offsetx - size.width() - brickwidth;
      if (invert) {
         currX += brickwidth;
      }
      while (currX < size.width()) {
         painter.drawRect(currX, currY, brickwidth, brickheight);
         currX += brickwidth * 2;
      }
      currY += brickheight;
   }
}

File: generators/checkboard.js
const generator = {
  apiVersion: 1,
  name: "Checkboard",
  description: "Draws a chequerboard pattern over an optional background.",
  type: "generator",
  inputs: ["Background"],

  // Each entry creates a control in the node settings panel.
  settings: [
    {
      id: "color",
      type: "color",
      name: "Colour",
      description: "Colour used for alternating squares.",
      default: { r: 0, g: 0, b: 0, a: 255 },
    },
    {
      id: "brickwidth",
      type: "real",
      name: "Square width (%)",
      description: "Width of each square as a percentage of the texture width.",
      default: 10,
      min: 1,
      max: 100,
    },
    {
      id: "brickheight",
      type: "real",
      name: "Square height (%)",
      description: "Height of each square as a percentage of the texture height.",
      default: 10,
      min: 1,
      max: 100,
    },
    {
      id: "offsetx",
      type: "integer",
      name: "Horizontal offset (%)",
      description: "Moves the pattern horizontally as a percentage of the texture width.",
      default: 0,
      min: -100,
      max: 100,
    },
    {
      id: "offsety",
      type: "integer",
      name: "Vertical offset (%)",
      description: "Moves the pattern vertically as a percentage of the texture height.",
      default: 0,
      min: -100,
      max: 100,
    },
  ],

  // This function runs whenever the application renders the node.
  generate(size, settings, output, inputs) {
    // Step 1: copy the optional background into the output. The coloured squares
    // will be painted over this image. Without a background, start transparent.
    const pixels = output.data;
    const background = inputs.Background?.data;
    if (background) pixels.set(background);
    else pixels.fill(0);

    // Step 2: convert percentages from the settings panel into pixel measurements.
    // A square width of 10% uses one tenth of the texture width, for example.
    const width = size.width;
    const height = size.height;
    const stride = output.stride;
    const squareWidth = Math.max(1, Math.trunc(settings.brickwidth * width / 100));
    const squareHeight = Math.max(1, Math.trunc(settings.brickheight * height / 100));
    const offsetX = Math.trunc(settings.offsetx * width / 100);
    const offsetY = Math.trunc(settings.offsety * height / 100);
    const colour = settings.color;
    if (colour.a === 0) return;

    // Alpha values are stored as bytes from 0 to 255. Dividing by 255 converts
    // them to fractions from 0 to 1, which are easier to use when blending colours.
    const colourAlpha = colour.a / 255;
    const remainingBackground = 1 - colourAlpha;

    // Step 3: use the offsets as the top-left origin of the repeating pattern.
    // Coordinates before this origin become negative, but Math.floor() still places
    // them in the correct square, so the pattern repeats in every direction.
    const patternOriginX = offsetX;
    const patternOriginY = offsetY;

    // Step 4: examine every output pixel. Dividing its distance from the pattern
    // origin by the square size tells us which row and column of squares it belongs to.
    for (let y = 0; y < height; ++y) {
      const squareRow = Math.floor((y - patternOriginY) / squareHeight);

      for (let x = 0; x < width; ++x) {
        const squareColumn = Math.floor((x - patternOriginX) / squareWidth);

        // Even row-and-column sums are coloured; odd sums are left unchanged.
        // Moving into the next row or column changes the sum and alternates the square.
        const isColouredSquare = (squareRow + squareColumn) % 2 === 0;
        if (!isColouredSquare) continue;

        // Each pixel contains four consecutive bytes: red, green, blue, and alpha.
        const pixelOffset = y * stride + x * 4;

        // Step 5: an opaque colour completely replaces the background pixel.
        if (colour.a === 255) {
          pixels[pixelOffset] = colour.r;
          pixels[pixelOffset + 1] = colour.g;
          pixels[pixelOffset + 2] = colour.b;
          pixels[pixelOffset + 3] = 255;
          continue;
        }

        // A transparent background has no visible colour to preserve, so the
        // translucent square can also be assigned directly.
        const backgroundAlpha = pixels[pixelOffset + 3] / 255;
        if (backgroundAlpha === 0) {
          pixels[pixelOffset] = colour.r;
          pixels[pixelOffset + 1] = colour.g;
          pixels[pixelOffset + 2] = colour.b;
          pixels[pixelOffset + 3] = colour.a;
          continue;
        }

        // For two translucent colours, retain the visible contribution from both.
        // The result is stored as ordinary, non-premultiplied RGBA.
        const resultAlpha = colourAlpha + backgroundAlpha * remainingBackground;
        pixels[pixelOffset] = Math.round(
          (colour.r * colourAlpha
            + pixels[pixelOffset] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixelOffset + 1] = Math.round(
          (colour.g * colourAlpha
            + pixels[pixelOffset + 1] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixelOffset + 2] = Math.round(
          (colour.b * colourAlpha
            + pixels[pixelOffset + 2] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixelOffset + 3] = Math.round(resultAlpha * 255);
      }
    }
  },
};
